ws.c

```diff
@@ -221,34 +226,54 @@ get_ws_frame(unsigned char *in_buffer, int buf_len, unsigned char **payload_ptr,
 	if (length_field <= 125) {
 		payload_len = length_field;
 	} else if (length_field == 126) { /* msglen is 16bit */
+		uint16_t tmp16;
 		if (buf_len < 4)
 			return INCOMPLETE_DATA;
-		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		memcpy(&tmp16, in_buffer + pos, 2);
+		payload_len = ntohs(tmp16);
 		pos += 2;
 	} else if (length_field == 127) { /* msglen is 64bit */
+		int i;
+		uint64_t tmp64 = 0;
 		if (buf_len < 10)
 			return INCOMPLETE_DATA;
-		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
-		pos += 8;
+		/* swap bytes from big endian to host byte order */
+		for (i = 56; i >= 0; i -= 8) {
+			tmp64 |= (uint64_t)in_buffer[pos++] << i;
```

Comment:
@cculianu sorry for the delay

What about this comment?

> But the code is wrong, (uint64_t)in_buffer[pos++] will read uint64 from the in_buffer always, while you need to read only one byte - (uint64_t)(data[pos++] & 0xff)

Reply:
Also can you please add a regression test for this, AFAIU it should be pretty easy, just use bigger messages

Reply:
Nevermind, the problem exists only if `in_buffer` was `char *`, it is ok for `unsigned char *`.

```
(gdb) p/x (char)0xde<<8
$4 = 0xffffde00
(gdb) p/x (unsigned char)0xde<<8
$5 = 0xde00

```